### PR TITLE
feat: accept function param in `initTRPC.context`

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

- Adds support for a function in `initTRPC.create` so we can do something like this:

```ts
// server/api/trpc.ts
export const createContext = () => { ... };

const t = initTRPC
  .context<typeof createContext>()
  .create();

export const publicProcedure = t.procedure;
// ...

// pages/api/trpc/[trpc].ts
import { createContext } from "~/server/api/trpc";

// ...
```

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/server/src/core/initTRPC.ts [hunk 1]
```diff
@@ -44,9 +44,20 @@ type CreateRootConfigTypesFromPartial<TTypes extends PartialRootConfigTypes> =
  * - Doesn't need to be a class but it doesn't really hurt either
  */
 
+// If the type is a function (sync or async) unwrap it, else use the type as is
+type Unwrap<TType> = TType extends (...args: any[]) => infer R
+  ? Awaited<R>
```
Comment (KATT) [inline packages/server/src/core/initTRPC.ts:49]: This requires TS 4.5 so we shouldn't do that one until next major

Comment (KATT) [reply packages/server/src/core/initTRPC.ts:49]: I want to avoid us changing reqs of TS versions unless it enables big features

Comment (juliusmarminge) [reply packages/server/src/core/initTRPC.ts]: used our custom `ThenArg` type for now

**Review by KATT — COMMENTED** [review]
Docs? :)

**Comment by juliusmarminge** [comment]
> Docs? :)

Done in https://github.com/trpc/trpc/pull/3495/commits/8d354563dff830b6209089394b7c1a2319112ab5

**Review by KATT — APPROVED** [review]

**Pull request merged**